fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/IndexingHeartbeatLowLevelTest.java

```diff
@@ -0,0 +1,473 @@
+        IndexingHeartbeat heartbeatMutual = new IndexingHeartbeat(UUID.randomUUID(), "Test", TimeUnit.SECONDS.toMillis(10), true);
+        IndexingHeartbeat heartbeatExclusive = new IndexingHeartbeat(UUID.randomUUID(), "Test", TimeUnit.SECONDS.toMillis(10), false);
+
+        // lock exclusive, then successfully lock mutual.
+        openSimpleMetaData(hook);
+        try (FDBRecordContext context = openContext()) {
+            heartbeatExclusive.checkAndUpdateHeartbeat(recordStore, index).join();
+            context.commit();
+        }
+
+        try (FDBRecordContext context = openContext()) {
+            heartbeatMutual.checkAndUpdateHeartbeat(recordStore, index).join();
+            // and clear
+            heartbeatExclusive.clearHeartbeat(recordStore, index);
+            heartbeatMutual.clearHeartbeat(recordStore, index);
+            context.commit();
+        }
+
+        // lock mutual, then fail to lock exclusive
+        try (FDBRecordContext context = openContext()) {
+            final Map<UUID, IndexBuildProto.IndexBuildHeartbeat> heartbeats = IndexingHeartbeat.getIndexingHeartbeats(recordStore, index, 0).join();
+            Assertions.assertThat(heartbeats).isEmpty();
+            heartbeatMutual.checkAndUpdateHeartbeat(recordStore, index).join();
+            context.commit();
+        }
+
+        try (FDBRecordContext context = openContext()) {
+            Assertions.assertThatThrownBy(() -> heartbeatExclusive.checkAndUpdateHeartbeat(recordStore, index).join())
+                    .isInstanceOf(CompletionException.class)
+                    .hasCauseInstanceOf(SynchronizedSessionLockedException.class);
+            // and clear
+            heartbeatMutual.clearHeartbeat(recordStore, index);
+            context.commit();
+        }
+
+    }
+
+    @Test
+    void testUnparseableHeartbeat() {
+        Index index = new Index("versionIndex1", concat(field("num_value_2"), VersionKeyExpression.VERSION), IndexTypes.VERSION);
+        FDBRecordStoreTestBase.RecordMetaDataHook hook = allIndexesHook(List.of(index));
+
+        // lock exclusive, then successfully lock mutual.
+        openSimpleMetaData(hook);
+        try (FDBRecordContext context = openContext()) {
+            byte[] key = IndexingSubspaces.indexHeartbeatSubspace(recordStore, index, UUID.randomUUID()).pack();
+            byte[] value = "meaningless byte value".getBytes();
+            recordStore.ensureContextActive().set(key, value);
+            context.commit();
+        }
+
+        IndexingHeartbeat heartbeat = new IndexingHeartbeat(UUID.randomUUID(), "Test", TimeUnit.SECONDS.toMillis(30), false);
+        try (FDBRecordContext context = openContext()) {
+            final Map<UUID, IndexBuildProto.IndexBuildHeartbeat> existingHeartbeats = IndexingHeartbeat.getIndexingHeartbeats(recordStore, index, 0).join();
+            Assertions.assertThat(existingHeartbeats).hasSize(1);
+            heartbeat.checkAndUpdateHeartbeat(recordStore, index);
```

Comment:
Probably would be good to use `clear` after this to clear both the unparseable heartbeat, and the new heartbeat.

Reply:
√